core/src/main/java/io/undertow/server/handlers/SameSiteCookieHandler.java

```diff
@@ -0,0 +1,139 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.undertow.server.handlers;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import io.undertow.server.HandlerWrapper;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.ResponseCommitListener;
+import io.undertow.server.handlers.builder.HandlerBuilder;
+import io.undertow.util.Headers;
+import io.undertow.util.SameSiteNoneIncompatibleClientChecker;
+
+/**
+ * Handler that will set the SameSite flag on all cookies
+ */
+public class SameSiteCookieHandler implements HttpHandler {
+
+    private final HttpHandler next;
+    private final String mode;
+    private final Pattern cookiePattern;
+    private final boolean disableClientChecker;
+
+    public SameSiteCookieHandler(final HttpHandler next, final String mode) {
+        this(next, mode, null, true, false);
+    }
+
+    public SameSiteCookieHandler(final HttpHandler next, final String mode, final boolean disableClientChecker) {
+        this(next, mode, null, true, disableClientChecker);
+    }
+
+    public SameSiteCookieHandler(final HttpHandler next, final String mode, final String cookiePattern) {
+        this(next, mode, cookiePattern, true, false);
+    }
+
+    public SameSiteCookieHandler(final HttpHandler next, final String mode, final String cookiePattern, final boolean caseSensitive) {
+        this(next, mode, cookiePattern, caseSensitive, false);
+    }
+
+    public SameSiteCookieHandler(final HttpHandler next, final String mode, final String cookiePattern, final boolean caseSensitive, final boolean disableClientChecker) {
+        this.next = next;
+        this.mode = mode;
+        if (cookiePattern != null && !cookiePattern.isEmpty()) {
+            this.cookiePattern = Pattern.compile(cookiePattern, caseSensitive ? 0 : Pattern.CASE_INSENSITIVE);
+        } else {
+            this.cookiePattern = null;
+        }
+        this.disableClientChecker = disableClientChecker;
+    }
+
+    @Override
+    public void handleRequest(HttpServerExchange exchange) throws Exception {
+        if (mode != null) {
+            exchange.addResponseCommitListener(new ResponseCommitListener() {
+                @Override
+                public void beforeCommit(HttpServerExchange exchange) {
+                    if (mode.equalsIgnoreCase(Cookie.SameSiteMode.None.toString())) {
```

Comment:
> For performance reasons, we need to change mode to the enum

Hmm..., "mode" is not the enum type at this stage because `Cookie#setSameSiteMode(String mode)` interface just accepts String. (It was my fault when I made the initial implementation... I understand the method should be better to accept the enum type, but we can not change the method due to compatibility reasons.)

Also, as `Cookie#setSameSiteMode(String mode)` can accept the "mode" with incase-sensitive, I think it's better to make `SameSiteCookieHandler` accept the "mode" with incase-sensitive. Therefore, as a user can specify the "mode" with incase-sensitive, we can not simply use `SameSiteMode.valueOf(mode)` to get the enum type from the "mode" value.
(`SameSiteMode.valueOf("None")` can return the enum type, but `SameSiteMode.valueOf("none")` throws `IllegalArgumentException`.) 


> and we should also verify if disableClientChecker is not true, right here, short-circuiting the check for mode if not needed.

Got it. From this suggestion, I think of another idea to avoid invoking the mode string comparison in every request.
 
I flip the flag name from "disableClientChecker" to "enableClientChecker". Then, As the client checker is required only for "SameSite=None",  "enableClientChecker" is set to true only if the mode is specified to "None" (incase-sensitive) in the constructor. The mode string comparison does not happen every request but only once in the constructor.